Add support for building repo with FreeBSD and Haiku

danboid opened the pull request:
I tweaked some makefiles and now the Uzebox repo buillds under FreeBSD (tested under v14) and Haiku (tested under beta4).

To build the repo:

FreeBSD / GhostBSD

pkg install gmake gcc sdl2 os-generic-userland-devtools avr-gcc avr-libc
<clone repo, cd uzebox>
gmake

Haiku

pkgman install avr_gcc avr_libc libsdl2_devel
<clone repo, cd uzebox>
make

EDIT 27th June 2024

This PR now also includes pcm2hex for converting audio files for use in Uzebox programs.

artcfox commented:
Please just put a dash in front of the original commands in the Makefile,
so they don't fail the build if it returns an error when people aren't
using the vendor-patched toolchains.

We shouldn't degrade the debug output for everyone, just because some
people want to use unofficial toolchains to build.

Adding a minus sign before the command name, like "-avr-size" should make
it ignore the error code when running that command.

And then people who run the official avr toolchains will be able to see
avr-specific output.



On Fri, Jun 21, 2024, 11:12 AM Dan MacDonald ***@***.***>
wrote:

> I tweaked some makefiles and now the Uzebox repo buillds under FreeBSD
> (tested under v14 )and Haiku (tested under beta4).
>
> To build the repo:
>
> FreeBSD / GhostBSD
>
> pkg install gmake gcc sdl2 os-generic-userland-devtools avr-gcc avr-libc
> <clone repo, cd uzebox>
> gmake
>
> Haiku
>
> pkgman install avr_gcc avr_libc libsdl2_devel
> <clone repo, cd uzebox>
> make
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Uzebox/uzebox/pull/134
> Commit Summary
>
>    - 8f1880a
>    <https://github.com/Uzebox/uzebox/pull/134/commits/8f1880a76f4d7dffc2b8359787c66545bb4603be>
>    Add support for FreeBSD
>    - 6f52fc8
>    <https://github.com/Uzebox/uzebox/pull/134/commits/6f52fc832a42eba7e028b0e52ea512c1fac41652>
>    Add support for Haiku
>
> File Changes
>
> (10 files <https://github.com/Uzebox/uzebox/pull/134/files>)
>
>    - *M* demos/BcDash/default/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-3c38c773a73e163468885a7afd712fb903fad36010588a90e43193e157fcbb8b>
>    (2)
>    - *M* demos/DonkeyKong/default/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-091696966d13d27bce85ff9ff4d2fe7bb38bfb9f271e5860bbb7240a5b84e3cb>
>    (8)
>    - *M* demos/Frogger/default/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-33deb6ca109d30df8d8c936462b9633a72cac32dfcbde303c4e95622c9dc185e>
>    (6)
>    - *M* demos/PacMan/default/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-756c52f429dab5c4c96733b48df1426efe6c6b970be56900f26df872dfd93712>
>    (2)
>    - *M* demos/SPIRamMusicDemo/default/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-2574b9faae66403cc3839ebb6a205f6218fe42c678a12341f929f69df48060f9>
>    (2)
>    - *M* demos/SpaceInvaders/default/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-4fbfdb500cf9953414d919f3bdf631ff82a0fc1e0c3563f4dae357671c247b2d>
>    (2)
>    - *M* demos/Zombienator/default/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-a4dabf90ee914a831964fb8f3041a6b053ec4af851f7f36dc33dce78f2206dee>
>    (2)
>    - *M* tools/gconvert/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-d50cd192064ccc2591cd95530770bbcbb6df363499d5b18e75cd45deae0e9a3c>
>    (20)
>    - *M* tools/packrom/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-51ebc4f44db29025ec84d1bbb929ed34f7f75e9d7f8e898e5f169065ee4a7d11>
>    (20)
>    - *M* tools/uzem/Makefile
>    <https://github.com/Uzebox/uzebox/pull/134/files#diff-3b48d0864c471355a847bfd5df4ba40868bbde8e4a4f1e4a39fa5c8393e2dfd9>
>    (17)
>
> Patch Links:
>
>    - https://github.com/Uzebox/uzebox/pull/134.patch
>    - https://github.com/Uzebox/uzebox/pull/134.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Uzebox/uzebox/pull/134>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAILLXY2UDPC4K7ZGDTICCTZIQ7HVAVCNFSM6AAAAABJWFAF2CVHI2DSMVQWIX3LMV43ASLTON2WKOZSGM3DMNZUGY4TONA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

danboid commented:
I can do that Matt but why is using `avr-size -A` instead of `avr-size -C` considered "degraded". `avr-size -A` is more compatible and seems to show more info about the binary than `avr-size -C` or am I missing something?

artcfox commented:
You could also use "-A --format=avr", but I suspect that is also not
supported by the unofficial toolchains.

The avr-specific output gives a much nicer output, showing program bytes
and data bytes added up along with a percentage full:

AVR Memory Usage
----------------
Device: atmega644

Program:   49650 bytes (75.8% Full)
(.text + .data + .bootloader)

Data:      3346 bytes (81.7% Full)
(.data + .bss + .noinit)

The official toolchains have a whole bunch of patches, that should probably
just be upstreamed. I only see binary downloads on Microchip's website, so
someone should probably get after them to actually release their source
files to not violate the GPL, and then we can upstream all the avr
enhancements.

On Fri, Jun 21, 2024, 11:26 AM Dan MacDonald ***@***.***>
wrote:

> I can do that Matt but why is using avr-size -A instead of avr-size -C
> considered "degraded". avr-size -A is more compatible and seems to show
> more info about the binary than avr-size -C or am I missing something?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Uzebox/uzebox/pull/134#issuecomment-2182966744>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAILLX3UZJYYBOMTIHCEWSTZIRA3XAVCNFSM6AAAAABJWFAF2CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCOBSHE3DMNZUGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

danboid commented:
I have updated my repo so that it reverts all of the `avr-size` commands to what they were but with a `-` in front and I can confirm that has fixed the builds and my updated repo builds fine under both FreeBSD 14 / GhostBSD and Haiku.

I have tested Uzem on both and it seems to work fine with sound on both platforms. I have not tested Uzem USB joystick support on either yet but I think it looks like it should be fully usable on both platforms now and I'd just use cuzebox anyway because that already builds under both FreeBSD and Haiku without any changes to the Makefile, Jubatian style.

danboid commented:
I expect it will be very easy to get the Uzebox repo building on netbsd  too because netbsd is pretty similar to FreeBSD and its also got avr-gcc and avr-libc packages but we'll leave this PR at just 2 new operating systems for today.

danboid commented:
I've updated the title and first post of this PR to note that this PR also adds the pcm2hex audio conversion tool.

danboid commented:
I have reverted my fork so that it only has the fixes for FreeBSD and Haiku.

@artcfox explained to me the audio format required for `bin2hex` to work so I have updated

https://uzebox.org/wiki/PCM_Sound_Effects

With the correct instructions on how to convert audio files for use with `bin2hex` which means `pcm2hex` isn't required.
